contrib/adapters/ale.py

```diff
@@ -0,0 +1,214 @@
+#    the License and to reproduce the content of the NOTICE file.
+#
+# You may obtain a copy of the Apache License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the Apache License with the above modification is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied. See the Apache License for the specific
+# language governing permissions and limitations under the Apache License.
+#
+
+"""OpenTimelineIO Avid Log Exchange (ALE) Adapter"""
+
+import opentimelineio as otio
+
+
+class ALEParseError(otio.exceptions.OTIOError):
+    pass
+
+
+def read_from_string(input_str, fps=24):
+
+    collection = otio.schema.SerializableCollection()
+    header = {}
+    columns = []
+
+    lines = input_str.split("\n")
+    while len(lines):
+        line = lines.pop(0)
+
+        # skip blank lines
+        if line.strip() == "":
+            continue
+
+        if line.strip() == "Heading":
+            while len(lines):
+                line = lines.pop(0)
+
+                if line.strip() == "":
+                    break
+
+                if "\t" not in line:
+                    raise ALEParseError("Invalid Heading line: "+line)
+
+                line = line.strip('\r')
+                segments = line.split("\t")
+                while len(segments) >= 2:
+                    key, val = segments.pop(0), segments.pop(0)
+                    header[key] = val
+                if len(segments) != 0:
+                    raise ALEParseError("Invalid Heading line: "+line)
+
+        if "FPS" in header:
+            fps = float(header["FPS"])
+
+        if line.strip() == "Column":
+            if len(lines) == 0:
+                raise ALEParseError("Unexpected end of file after: "+line)
+
+            line = lines.pop(0)
+            columns = line.split("\t")
+
+        if line.strip() == "Data":
+            while len(lines):
+                line = lines.pop(0)
+
+                if line.strip() == "":
+                    continue
+
+                row = line.split("\t")
+
+                if len(row) < len(columns):
+                    raise ALEParseError("Too few values on row: "+line)
+
+                if len(row) > len(columns):
+                    raise ALEParseError("Too many values on row: "+line)
+
+                metadata = dict(zip(columns, row))
+
+                clip = otio.schema.Clip()
+                clip.name = metadata.get("Name")
+                del metadata["Name"]
+
+                if "Start" in metadata:
+                    start = otio.opentime.from_timecode(metadata["Start"], fps)
+                    del metadata["Start"]
+                    duration = None
+                    end = None
+                    if "Duration" in metadata:
+                        duration = otio.opentime.from_timecode(
```

Comment:
I wrapped each timecode parse in a try/catch in case of blanks or malformed values, so they are skipped. I *could* make the blank a special case and complain about malformed values, that might be a better strategy...

Reply:
General note is that we haven't really built any kind of formal error/warning system into the adapters yet.  Still not sure exactly what the needs are there.